README.md
# Just my version of 2048 game


CHANGELOG.md
# Changelog

## Release 1.0.0 (2022-03-06)
* Basic working game (moving, merging)


tests.py
import unittest

import game2048


G = game2048.G
G.debug_headless = True
G.size = 4


class TestMoving(unittest.TestCase):
    def test_1(self):
        G.M = [[2, 8, 8, 0], [0, 16, 64, 8], [128, 32, 0, 4], [0, 8, 0, 64]]
        G.move_matrix(0, -1)
        self.assertEqual([[2, 8, 8, 8], [128, 16, 64, 4], [0, 32, 0, 64], [0, 8, 0, 0]], G.M)
        G.move_matrix(1, 0)
        self.assertEqual([[0, 2, 8, 16], [128, 16, 64, 4], [0, 0, 32, 64], [0, 0, 0, 8]], G.M)
        G.move_matrix(1, 0)
        self.assertEqual([[0, 2, 8, 16], [128, 16, 64, 4], [0, 0, 32, 64], [0, 0, 0, 8]], G.M)

    def test_2(self):
        G.M = [[2, 8, 8, 0], [0, 16, 64, 8], [128, 32, 0, 4], [0, 8, 0, 64]]
        G.move_matrix(1, 0)
        self.assertEqual([[0, 0, 2, 16], [0, 16, 64, 8], [0, 128, 32, 4], [0, 0, 8, 64]], G.M)

    def test_3(self):
        G.M = [[2, 8, 8, 0], [0, 16, 64, 8], [128, 32, 0, 4], [0, 8, 0, 64]]
        G.move_matrix(0, -1)
        self.assertEqual([[2, 8, 8, 8], [128, 16, 64, 4], [0, 32, 0, 64], [0, 8, 0, 0]], G.M)
        G.move_matrix(-1, 0)
        self.assertEqual([[2, 16, 8, 0], [128, 16, 64, 4], [32, 64, 0, 0], [8, 0, 0, 0]], G.M)
        G.move_matrix(0, 1)
        self.assertEqual([[2, 0, 0, 0], [128, 0, 0, 0], [32, 32, 8, 0], [8, 64, 64, 4]], G.M)
        G.move_matrix(-1, 0)
        self.assertEqual([[2, 0, 0, 0], [128, 0, 0, 0], [64, 8, 0, 0], [8, 128, 4, 0]], G.M)
        G.move_matrix(0, -1)
        self.assertEqual([[2, 8, 4, 0], [128, 128, 0, 0], [64, 0, 0, 0], [8, 0, 0, 0]], G.M)
        G.move_matrix(-1, 0)
        self.assertEqual([[2, 8, 4, 0], [256, 0, 0, 0], [64, 0, 0, 0], [8, 0, 0, 0]], G.M)

    def test_4(self):
        G.M = [[2, 2, 2, 2] for _ in range(4)]
        G.move_matrix(0, 1)
        self.assertEqual([[0, 0, 0, 0], [0, 0, 0, 0], [4, 4, 4, 4], [4, 4, 4, 4]], G.M)
        G.move_matrix(0, -1)
        self.assertEqual([[8, 8, 8, 8], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], G.M)
        G.move_matrix(-1, 0)
        self.assertEqual([[16, 16, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], G.M)
        G.move_matrix(1, 0)
        self.assertEqual([[0, 0, 0, 32], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], G.M)

    def test_5(self):
        G.M = [[2, 2, 0, 4], [8, 8, 16, 0], [0, 0, 0, 0], [0, 0, 0, 0]]
        G.move_matrix(-1, 0)
        self.assertEqual([[4, 4, 0, 0], [16, 16, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], G.M)
        G.move_matrix(-1, 0)
        self.assertEqual([[8, 0, 0, 0], [32, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], G.M)

    def test_6(self):
        G.M = [[16, 0, 0, 4], [0, 0, 0, 4], [16, 0, 0, 0], [32, 0, 0, 8]]
        G.move_matrix(0, 1)
        self.assertEqual([[0, 0, 0, 0], [0, 0, 0, 0], [32, 0, 0, 8], [32, 0, 0, 8]], G.M)
        G.move_matrix(0, -1)
        self.assertEqual([[64, 0, 0, 16], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], G.M)

    def test_7(self):
        G.M = [[2, 0, 0, 0], [2, 0, 0, 0], [4, 0, 0, 0], [4, 0, 0, 0]]
        G.move_matrix(0, -1)
        self.assertEqual([[4, 0, 0, 0], [8, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], G.M)

    def test_8(self):
        G.M = self.M = [[4, 0, 0, 0], [4, 0, 0, 0], [2, 0, 0, 0], [2, 0, 0, 0]]
        G.move_matrix(0, 1)
        self.assertEqual([[0, 0, 0, 0], [0, 0, 0, 0], [8, 0, 0, 0], [4, 0, 0, 0]], G.M)

    def test_9(self):
        G.M = self.M = [[4, 0, 0, 0], [4, 0, 0, 0], [2, 0, 0, 0], [2, 0, 0, 0]]
        G.move_matrix(0, -1)
        self.assertEqual([[8, 0, 0, 0], [4, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], G.M)

    def test_10(self):
        G.M = self.M = self.M = [[4, 4, 2, 2], [0, 0, 0, 0], [2, 2, 4, 4], [0, 0, 0, 0]]
        G.move_matrix(1, 0)
        self.assertEqual([[0, 0, 8, 4], [0, 0, 0, 0], [0, 0, 4, 8], [0, 0, 0, 0]], G.M)

    def test_11(self):
        G.M = self.M = self.M = [[4, 4, 2, 2], [0, 0, 0, 0], [2, 2, 4, 4], [0, 0, 0, 0]]
        G.move_matrix(-1, 0)
        self.assertEqual([[8, 4, 0, 0], [0, 0, 0, 0], [4, 8, 0, 0], [0, 0, 0, 0]], G.M)

    def test_12(self):
        G.M = [[8, 8, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]
        G.move_matrix(1, 0)
        self.assertEqual([[0, 0, 0, 16], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], G.M)


run_tests.sh
#!/bin/bash

path=$(dirname "$0")
cd $path
/usr/bin/env python3 -m unittest tests.py


2048.py
#!/usr/bin/env python3

import pygame


class Text:
    def __init__(self, text, color):
        self.text_width, self.text_height = font.size(text)
        self.text = font.render(text, True, color)

    def draw(self, x, y):
        window.blit(self.text, (x, y))


class Cell:
    def __init__(self, value):
        self.value = value
        self.color = cell_colors[value]
        if value:
            self.text = Text(str(value), font_colors[value])

    def draw(self, row, col):
        pygame.draw.rect(window, self.color, (col * cell_size, row * cell_size, cell_size, cell_size))
        if self.value:
            self.text.draw(col * cell_size + (cell_size - self.text.text_width) / 2, row * cell_size + (cell_size - self.text.text_height) / 2)


def merge_cells(row_absorbent, col_absorbent, row_absorbed, col_absorbed):
    Matrix[row_absorbent][col_absorbent] *= 2
    Matrix[row_absorbed][col_absorbed] = 0


def find_maximum_movement(row_start, col_start, dir_x, dir_y):
    if dir_x != 0:  # x direction
        col_end = col_start + dir_x
        while 0 < col_end < 3 and not Matrix[row_start][col_end] and ((not Matrix[row_start][col_end + dir_x]) or Matrix[row_start][col_start] == Matrix[row_start][col_end + dir_x]):
            col_end += dir_x
        return col_end
    else:  # y direction
        row_end = row_start + dir_y
        while 0 < row_end < 3 and not Matrix[row_end][col_start] and ((not Matrix[row_end + dir_y][col_start]) or Matrix[row_start][col_start] == Matrix[row_end + dir_y][col_start]):
            row_end += dir_y
        return row_end


def move_cells(dir_x, dir_y):
    if dir_x != 0:  # x direction
        for row in range(0, 4):
            for col_start in (range(2, -1, -1) if dir_x > 0 else range(0, 4)):
                if not ((col_start == 0 and dir_x < 0) or (col_start == 3 and dir_x > 0)):
                    col_end = find_maximum_movement(row, col_start, dir_x, dir_y)
                    if col_start != col_end:
                        if not Matrix[row][col_end]:  # is empty
                            Matrix[row][col_end] = Matrix[row][col_start]
                            Matrix[row][col_start] = 0
                        elif Matrix[row][col_start] == Matrix[row][col_end]:
                            merge_cells(row, col_end, row, col_start)
    else:  # y direction
        for col in range(0, 4):
            for row_start in (range(2, -1, -1) if dir_y > 0 else range(0, 4)):
                if not ((row_start == 0 and dir_y < 0) or (row_start == 3 and dir_y > 0)):
                    row_end = find_maximum_movement(row_start, col, dir_x, dir_y)
                    if row_start != row_end:
                        if not Matrix[row_end][col]:  # is empty
                            Matrix[row_end][col] = Matrix[row_start][col]
                            Matrix[row_start][col] = 0
                        elif Matrix[row_start][col] == Matrix[row_end][col]:
                            merge_cells(row_end, col, row_start, col)


def main_loop():
    game_notOver = True
    move_time_remaining = 0
    dir_x, dir_y = 0, 0
    while game_notOver:
        clock.tick(fps)

        keys = pygame.key.get_pressed()
        mouse = pygame.mouse.get_pos()

        for event in pygame.event.get():
            if event.type == pygame.QUIT or keys[pygame.K_ESCAPE]:
                game_notOver = False

        if not move_time_remaining:
            if keys[pygame.K_LEFT] or keys[pygame.K_a]:
                move_time_remaining = move_time
                dir_x = -1
                dir_y = 0
            if keys[pygame.K_RIGHT] or keys[pygame.K_d]:
                move_time_remaining = move_time
                dir_x = 1
                dir_y = 0
            if keys[pygame.K_UP] or keys[pygame.K_w]:
                move_time_remaining = move_time
                dir_x = 0
                dir_y = -1
            if keys[pygame.K_DOWN] or keys[pygame.K_s]:
                move_time_remaining = move_time
                dir_x = 0
                dir_y = 1

            if move_time_remaining:
                move_cells(dir_x, dir_y)
        else:
            move_time_remaining -= 1

        redraw()


def redraw():
    window.fill((250, 248, 239))

    for row, row_of_cells in enumerate(Matrix):
        for col, val in enumerate(row_of_cells):
            Cells[val].draw(row, col)

    pygame.display.update()


cell_size = 100
fps = 60
move_time = fps * 1

values = [0] + [2 ** i for i in range(1, 7 + 1)]

cell_colors = {
    0: (205, 193, 180),
    2: (238, 228, 218),
    4: (238, 225, 201),
    8: (243, 178, 122),
    16: (246, 150, 100),
    32: (247, 124, 95),
    64: (247, 95, 59),
    128: (237, 208, 115),
    256: (237, 204, 98),
    512: (237, 201, 80),
    1024: (237, 197, 63),
    2048: (237, 194, 46),
}

font_colors = {
    2: (119, 110, 101),
    4: (119, 110, 101),
    8: (249, 246, 242),
    16: (249, 246, 242),
    32: (249, 246, 242),
    64: (249, 246, 242),
    128: (249, 246, 242),
    256: (249, 246, 242),
    512: (249, 246, 242),
    1024: (249, 246, 242),
    2048: (249, 246, 242),
}

Matrix = [[2, 8, 8, 0], [0, 16, 64, 8], [128, 32, 0, 4], [0, 8, 0, 64]]


if __name__ == "__main__":
    pygame.display.init()
    pygame.font.init()
    clock = pygame.time.Clock()
    window = pygame.display.set_mode((400, 400))

    font = pygame.font.SysFont("Verdana", 40, bold=True)
    Cells = {i: Cell(i) for i in values}

    main_loop()

    pygame.quit()


tests-anim.py
import unittest

import game2048

G = game2048.G
G.debug_anim = True
G.size = 4


class TestAnimation(unittest.TestCase):
    def test_1(self):
        print("\n↑")
        G.M = [[0, 0, 0, 0], [16, 0, 0, 0], [4, 0, 0, 0], [4, 0, 0, 0]]
        G.start()
        self.assertTrue(G.debug_exit_status)

    def test_2(self):
        print("\n↓")
        G.M = [[4, 0, 0, 0], [4, 0, 0, 0], [2, 0, 0, 0], [2, 0, 0, 0]]
        G.start()
        self.assertTrue(G.debug_exit_status)

    def test_3(self):
        print("\n↑")
        G.M = [[2, 0, 0, 0], [2, 0, 0, 0], [4, 0, 0, 0], [4, 0, 0, 0]]
        G.start()
        self.assertTrue(G.debug_exit_status)

    def test_4(self):
        print("\n→")
        G.M = [[4, 4, 2, 2], [0, 0, 0, 0], [2, 2, 4, 4], [0, 0, 0, 0]]
        G.start()
        self.assertTrue(G.debug_exit_status)

    def test_5(self):
        print("\n←")
        G.M = [[4, 4, 2, 2], [0, 0, 0, 0], [2, 2, 4, 4], [0, 0, 0, 0]]
        G.start()
        self.assertTrue(G.debug_exit_status)


game2048.py
#!/usr/bin/env python3

import os
import pygame
import random


class Colors:
    background = (250, 248, 239)
    board_background = (187, 173, 160)
    cell = {
        0: (205, 193, 180),
        2: (238, 228, 218),
        4: (238, 225, 201),
        8: (243, 178, 122),
        16: (246, 150, 100),
        32: (247, 124, 95),
        64: (247, 95, 59),
        128: (237, 208, 115),
        256: (237, 204, 98),
        512: (237, 201, 80),
        1024: (237, 197, 63),
        2048: (237, 194, 46),
    }
    font = {
        2: (119, 110, 101),
        4: (119, 110, 101),
        8: (249, 246, 242),
        16: (249, 246, 242),
        32: (249, 246, 242),
        64: (249, 246, 242),
        128: (249, 246, 242),
        256: (249, 246, 242),
        512: (249, 246, 242),
        1024: (249, 246, 242),
        2048: (249, 246, 242),
    }


class Text:
    def __init__(self, text, color):
        self.text_width, self.text_height = G.font.size(text)
        self.text = G.font.render(text, True, color)

    def draw(self, x, y):
        G.window.blit(self.text, (x, y))


class RoundedSquare:
    def __init__(self, a, r, color):
        self.a = a
        self.r = r
        self.color = color

    def draw(self, x, y):
        pygame.draw.rect(G.window, self.color, (x + self.r, y, self.a - 2 * self.r, self.a))
        pygame.draw.rect(G.window, self.color, (x, y + self.r, self.a, self.a - 2 * self.r))
        pygame.draw.circle(G.window, self.color, (x + self.r, y + self.r), self.r)
        pygame.draw.circle(G.window, self.color, (x + self.a - self.r, y + self.r), self.r)
        pygame.draw.circle(G.window, self.color, (x + self.a - self.r, y + self.a - self.r), self.r)
        pygame.draw.circle(G.window, self.color, (x + self.r, y + self.a - self.r), self.r)


class Cell:
    def __init__(self, value):
        self.value = value
        self.tile = RoundedSquare(G.cell_size - 2 * G.grid_spacing, G.radius, Colors.cell[value])
        if value:
            self.text = Text(str(value), Colors.font[value])

    def draw(self, x_real, y_real):
        self.tile.draw(x_real + G.grid_spacing, y_real + G.grid_spacing)
        if self.value:
            self.text.draw(x_real + (G.cell_size - self.text.text_width) / 2, y_real + (G.cell_size - self.text.text_height) / 2)


class Animation:
    def __init__(self, row_start, col_start, row_end, col_end, value):
        self.value = value
        self.row_start = row_start
        self.col_start = col_start
        self.row_end = row_end
        self.col_end = col_end
        self.move_distance_x = (col_end - col_start) * G.cell_size
        self.move_distance_y = (row_end - row_start) * G.cell_size
        self.move_step_x = self.move_distance_x // G.move_time
        self.move_step_y = self.move_distance_y // G.move_time
        self.start_pos_x = col_start * G.cell_size
        self.start_pos_y = row_start * G.cell_size

    def draw(self):
        pos_now_x = self.start_pos_x + self.move_step_x * (G.move_time - G.move_time_remaining)
        pos_now_y = self.start_pos_y + self.move_step_y * (G.move_time - G.move_time_remaining)

        G.CellsPrerendered[self.value].draw(pos_now_x, pos_now_y)

    def __repr__(self):
        return f"Anim(start=({self.row_start}, {self.col_start}, dist=({self.col_end - self.col_start}, {self.row_end - self.row_start}))"


class Game:
    debug_headless = False
    debug_anim = False
    debug_exit_status = None

    size = 4
    cell_size = 120
    radius = 9
    grid_spacing = 6
    fps = 60
    values = [0] + [2 ** i for i in range(1, 11 + 1)]

    def __init__(self):
        self.M = [[0 for _ in range(self.size)] for _ in range(self.size)]  # Matrix
        self.spawn_cell()
        self.spawn_cell()
        self.reset_animations()

    def start(self):
        self.move_time = 60 if self.debug_anim else 20
        self.reset_animations()
        pygame.display.init()
        pygame.font.init()
        self.clock = pygame.time.Clock()
        self.window = pygame.display.set_mode((self.size * self.cell_size, self.size * self.cell_size), vsync=1)
        pygame.display.set_caption("2048")
        self.font = pygame.font.Font("OpenSans-Bold.ttf", 52)
        self.CellsPrerendered = {i: Cell(i) for i in self.values}

        self.main_loop()

        pygame.quit()

    def main_loop(self):
        self.game_notOver = True
        self.move_time_remaining = 0
        dir_x, dir_y = 0, 0
        while self.game_notOver:
            self.clock.tick(self.fps)

            keys = pygame.key.get_pressed()
            # mouse = pygame.mouse.get_pos()

            for event in pygame.event.get():
                if event.type == pygame.QUIT or keys[pygame.K_ESCAPE]:
                    self.game_notOver = False

            if self.move_time_remaining == 0:
                if self.cells_being_animated:  # something has moved
                    self.reset_animations()
                    self.spawn_cell()

                pressed_key = True
                if keys[pygame.K_LEFT] or keys[pygame.K_a]:  # ←
                    dir_x, dir_y = -1, 0
                elif keys[pygame.K_RIGHT] or keys[pygame.K_d]:  # →
                    dir_x, dir_y = 1, 0
                elif keys[pygame.K_UP] or keys[pygame.K_w]:  # ↑
                    dir_x, dir_y = 0, -1
                elif keys[pygame.K_DOWN] or keys[pygame.K_s]:  # ↓
                    dir_x, dir_y = 0, 1
                else:
                    pressed_key = False

                if pressed_key:
                    self.move_time_remaining = self.move_time
                    self.move_matrix(dir_x, dir_y)

                if self.debug_anim and (keys[pygame.K_y] or keys[pygame.K_n]):
                    self.debug_exit_status = keys[pygame.K_y]
                    self.game_notOver = False
            else:
                self.move_time_remaining -= 1

            self.redraw()

    def redraw(self):
        G.window.fill(Colors.board_background)

        for row, row_of_cells in enumerate(self.M):
            for col, val in enumerate(row_of_cells):
                if (row, col) in self.cells_being_animated:
                    G.CellsPrerendered[0].draw(col * self.cell_size, row * self.cell_size)
                else:
                    G.CellsPrerendered[val].draw(col * self.cell_size, row * self.cell_size)

        for anim in self.animations:
            anim.draw()

        pygame.display.update()

    def spawn_cell(self):
        if not self.debug_anim:
            free_tiles = [(row, col) for row, row_of_cells in enumerate(self.M) for col, val in enumerate(row_of_cells) if val == 0]
            row, col = random.choice(free_tiles)
            self.M[row][col] = random.choice([2, 2, 2, 2, 2, 2, 2, 2, 2, 4])

    def reset_animations(self):
        self.animations = set()
        self.cells_being_animated = set()

    def add_animation(self, row_start, col_start, row_end, col_end, value):
        if not self.debug_headless:
            self.animations.add(Animation(row_start, col_start, row_end, col_end, value))
            self.cells_being_animated.add((row_end, col_end))

    def move_cell(self, row_start, col_start, row_end, col_end):
        self.add_animation(row_start, col_start, row_end, col_end, self.M[row_start][col_start])
        self.M[row_end][col_end] = self.M[row_start][col_start]
        self.M[row_start][col_start] = 0

    def merge_cells(self, row_absorbent, col_absorbent, row_absorbed, col_absorbed):
        if not self.debug_headless:
            # ↓ absorbent animation, check if absorbent is not already being moved
            if (row_absorbent, col_absorbent) not in self.cells_being_animated:
                self.add_animation(row_absorbent, col_absorbent, row_absorbent, col_absorbent, self.M[row_absorbent][col_absorbent])
            # ↓ absorbed animation
            self.add_animation(row_absorbed, col_absorbed, row_absorbent, col_absorbent, self.M[row_absorbed][col_absorbed])

        self.M[row_absorbent][col_absorbent] *= 2
        self.M[row_absorbed][col_absorbed] = 0
        self.no_longer_mergeable.add((row_absorbent, col_absorbent))

    def find_maximum_movement(self, row_start, col_start, dir_x, dir_y):
        if dir_x != 0:  # x direction
            col_end = col_start + dir_x
            while 0 < col_end < self.size - 1 and self.M[row_start][col_end] == 0 and (self.M[row_start][col_end + dir_x] == 0 or self.M[row_start][col_start] == self.M[row_start][col_end + dir_x]) and (row_start, col_end + dir_x) not in self.no_longer_mergeable:
                col_end += dir_x
            return col_end
        else:  # y direction
            row_end = row_start + dir_y
            while 0 < row_end < self.size - 1 and self.M[row_end][col_start] == 0 and (self.M[row_end + dir_y][col_start] == 0 or self.M[row_start][col_start] == self.M[row_end + dir_y][col_start]) and (row_end + dir_y, col_start) not in self.no_longer_mergeable:
                row_end += dir_y
            return row_end

    def move_matrix(self, dir_x, dir_y):
        self.no_longer_mergeable = set()
        if dir_x != 0:  # x direction
            for row in range(0, self.size):
                for col_start in (range(self.size - 2, -1, -1) if dir_x > 0 else range(0, self.size)):
                    if self.M[row][col_start] and not ((col_start == 0 and dir_x < 0) or (col_start == self.size - 1 and dir_x > 0)):
                        col_end = self.find_maximum_movement(row, col_start, dir_x, dir_y)
                        if self.M[row][col_end] == 0:
                            self.move_cell(row, col_start, row, col_end)
                        elif self.M[row][col_start] == self.M[row][col_end] and (row, col_end) not in self.no_longer_mergeable:
                            self.merge_cells(row, col_end, row, col_start)
        else:  # y direction
            for col in range(0, self.size):
                for row_start in (range(self.size - 2, -1, -1) if dir_y > 0 else range(0, self.size)):
                    if self.M[row_start][col] and not ((row_start == 0 and dir_y < 0) or (row_start == self.size - 1 and dir_y > 0)):
                        row_end = self.find_maximum_movement(row_start, col, dir_x, dir_y)
                        if self.M[row_end][col] == 0:
                            self.move_cell(row_start, col, row_end, col)
                        elif self.M[row_start][col] == self.M[row_end][col] and (row_end, col) not in self.no_longer_mergeable:
                            self.merge_cells(row_end, col, row_start, col)


os.chdir(os.path.dirname(__file__))

G = Game()
if __name__ == "__main__":
    G.start()
